### PR TITLE
Docker

### PR DESCRIPTION
**Pull request by VolkerHartmann**
First version supporting docker.

### CODE REVIEW

File: docker/buildDocker.sh [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+################################################################################
+
+################################################################################
+# Test for commands used in this script
+################################################################################
+testForCommands="dirname date docker git"
+
+for command in $testForCommands
+do 
+  $command --help >> /dev/null
```
Comment (Pfeil) [inline docker/buildDocker.sh:44]: Not sure about linux, but on OSX (10.15.7) "--help" is not available for `dirname` and `date`. I think `type $command` could work more cross platform? Not sure what the most accepted solution is.

```suggestion
  type $command >> /dev/null
```

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+# Create run script
+################################################################################
+printInfo Create run script ...
+
+# Determine name of jar file.
+for file in $INSTALLATION_DIRECTORY/*.jar; do
+  jarFile=${file##*/}
+done
+
+echo #!/bin/bash                                                                                >  $INSTALLATION_DIRECTORY/run.sh
```
Comment (Pfeil) [inline build.sh:142]: Every line beginning with "echo #" creates just an empty line in the output of build.sh. You need to escape: # -> \# or maybe just make ordinary script comments that will not be transferred into the run.sh. Not sure what is better.

File: build.sh [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+mkdir $INSTALLATION_DIRECTORY/lib
+
+###############################################################################
+# Create run script
+################################################################################
+printInfo Create run script ...
+
+# Determine name of jar file.
+for file in $INSTALLATION_DIRECTORY/*.jar; do
+  jarFile=${file##*/}
```
Comment (Pfeil) [inline build.sh:139]: Not exactly sure how this works (bash is magic) but the variable does not seem to work for me, as my script is generated like this:

```bash
jarFile=*.jar
ACTUAL_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
cd "$ACTUAL_DIR"
java -cp ".:$jarFile" -Dloader.path="file://$ACTUAL_DIR/$jarFile,./lib/,." -jar $jarFile
```

And after starting, the container output is: `Error: Unable to access jarfile *.jar`

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ARG REPO_PORT_DEFAULT=8040
+ARG SERVICE_ROOT_DIRECTORY_DEFAULT=/spring/
+####################################################
+# END GLOBAL DECLARATION
+####################################################
+
+####################################################
+# Building environment (java & git)
+####################################################
+FROM openjdk:11-stretch AS build-env-java
```
Comment (Pfeil) [inline Dockerfile:14]: I am successfully using `openjdk:11-buster` for the PIT service and the indexer within the testbed. It's just the base-os version, but maybe as we are at it anyway, should we update?

File: build.gradle [hunk 5]
```diff
@@ -69,7 +69,12 @@ dependencies {
     implementation "org.springdoc:springdoc-openapi-data-rest:1.2.34"
     implementation "org.springdoc:springdoc-openapi-webmvc-core:1.2.34"
 
+    // driver for postgres
     implementation "org.postgresql:postgresql:42.2.1"
+    //driver for h2
+    implementation "com.h2database:h2:1.4.200"
```
Comment (Pfeil) [inline build.gradle:75]: I know this was mentioned in the chat, but I do still not understand what makes this change necessary, sorry :D Is this a worksaround for something?

**Review by Pfeil — CHANGES_REQUESTED** [review]
I like how generic it is, this will be useful in the future. But it does not yet work for me.

I noted what my IDE told me, and also tried to build the image and run the container. I managed to work around the issue with `dirname` and `date` as mentioned, but ran into the issue with the run.sh, so the container started and then exited immediately as mentioned.

File: build.sh [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+################################################################################
+
+################################################################################
+# Test for commands used in this script
+################################################################################
+testForCommands="dirname git date java javac"
+
+for command in $testForCommands
+do 
+  $command --help >> /dev/null
```
[the remark this answers is not in the document]
Comment (VolkerHartmann) [reply build.sh:86]: Your suggestion also works under Linux. So this should at least be a better solution.

File: Dockerfile (hunk 4 above)
Comment (VolkerHartmann) [reply Dockerfile:14]: :+1:  We should use the same base everywhere. Then we have it easier to keep the updates up to date.

File: build.gradle (hunk 5 above)
Comment (VolkerHartmann) [reply build.gradle:75]: For those who want to use the docker container out-of-the-box without additional overhead h2 is the default database. For productive use the configuration can be switched to postgres without any additional effort.

File: build.sh (hunk 2 above)
Comment (VolkerHartmann) [reply build.sh:142]: echo "...." solves the problem

File: build.sh (hunk 3 above)
Comment (VolkerHartmann) [reply build.sh:139]: Ok due to the problems I found an easier solution:
cd $INSTALLATION_DIRECTORY
# Determine name of jar file.
jarFile=(`ls $REPO_NAME*.jar`)

should solve the problem.

File: build.sh [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+################################################################################
+# Check parameters
+################################################################################
+checkParameters $*
+
+################################################################################
+# Determine repo name 
+################################################################################
+REPO_URL=`git remote get-url origin`
+REPO_NAME=${REPO_URL##*/}
```
Comment (Pfeil) [inline build.sh:107]: If a repository was cloned with it's .git suffix, it must be removed to find the jar file later on. With the following change, I was then able to run the container and access the swagger interface successfully:

```suggestion
REPO_NAME=${REPO_URL##*/}
REPO_NAME=${REPO_NAME%".git"}
```

File: build.sh [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+# Create run script
+################################################################################
+printInfo Create run script ...
+
+cd $INSTALLATION_DIRECTORY
+
+# Determine name of jar file.
+jarFile=(`ls $REPO_NAME*.jar`)
+
+echo "#!/bin/bash"                                                                              >  $INSTALLATION_DIRECTORY/run.sh
```
Comment (Pfeil) [inline build.sh:142]: So, when building locally on OSX using `bash build.sh installdir` I do get those errors, and I do not have a fix for that as I do not fully understand the problem yet:

```
---------------------------------------------------------------------------
Create run script ...
---------------------------------------------------------------------------
build.sh: line 143: installme/run.sh: No such file or directory
build.sh: line 144: installme/run.sh: No such file or directory
build.sh: line 145: installme/run.sh: No such file or directory
build.sh: line 146: installme/run.sh: No such file or directory
build.sh: line 147: installme/run.sh: No such file or directory
build.sh: line 148: installme/run.sh: No such file or directory
build.sh: line 149: installme/run.sh: No such file or directory
build.sh: line 150: installme/run.sh: No such file or directory
build.sh: line 151: installme/run.sh: No such file or directory
build.sh: line 152: installme/run.sh: No such file or directory
build.sh: line 153: installme/run.sh: No such file or directory
build.sh: line 154: installme/run.sh: No such file or directory
build.sh: line 155: installme/run.sh: No such file or directory
build.sh: line 156: installme/run.sh: No such file or directory
chmod: installme/run.sh: No such file or directory
.
---------------------------------------------------------------------------
Now you can start the service by calling installme/run.sh
---------------------------------------------------------------------------
```

File: build.sh [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    usage
+    exit 1
+  fi
+  
+  # Check for invalid flags
+  if [ "${1:0:1}" = "-" ]; then
+    usage
+  fi
+
+  INSTALLATION_DIRECTORY=$1
```
Comment (Pfeil) [inline build.sh:48]: As autocompletion may add a / as a suffix to the given folder name, should we remove it in this case?

```suggestion
  INSTALLATION_DIRECTORY=$1
  INSTALLATION_DIRECTORY=${INSTALLATION_DIRECTORY%"/"}
```

**Review by Pfeil — CHANGES_REQUESTED** [review]
So with only minor changes the docker build worked for me :) I am a bit confused about the problem with the local build though. It's probably something simple, but I was not able to fully reproduce the problem outside of the script. I think that some of the quoted strings maybe still get interpreted before writing it into the file.

File: build.sh (hunk 8 above)
Comment (VolkerHartmann) [reply build.sh]: Ahrg. I always tried with absolute paths which works fine.

File: build.sh (hunk 9 above)
Comment (VolkerHartmann) [reply build.sh:48]: Relative paths are always a problem. I transfer the installation directory to an absolute one. So this is no longer necessary.

**Review by Pfeil — APPROVED** [review]
Works for me 👍  (both as a container and building locally).

**Comment by VolkerHartmann** [comment]
Close #5

**Pull request merged**